src/main/java/jenkins/advancedqueue/sorter/StartedJobItemCache.java

```diff
@@ -0,0 +1,153 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.advancedqueue.sorter;
+
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.concurrent.TimeUnit;
+
+import com.google.common.base.Objects;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+
+import hudson.model.Run;
+import hudson.model.Queue.Executable;
+import hudson.model.queue.WorkUnit;
+
+/**
+ * Keeps track of the Queue.Items seen by the Sorter, but removed from the queue
+ * to become jobs, for UpstreamCauseStrategy.
+ *
+ * @author Ronny Schuetz
+ * @since 2.3
+ */
+public class StartedJobItemCache {
+
+	private static final int RETENTION_COUNT = 10000;
+	private static final int RETENTION_TIME_HOURS = 12;
+
+	private static StartedJobItemCache startedJobItemCache = null;
+
+	static {
+		startedJobItemCache = new StartedJobItemCache();
+	}
+
+	public static StartedJobItemCache get() {
+		return startedJobItemCache;
+	}
+
+	private static class PendingItem {
+		final long startTime;
+		final ItemInfo itemInfo;
+		final WorkUnit workUnit;
+
+		public PendingItem(final ItemInfo itemInfo, final WorkUnit workUnit) {
+			this.startTime = System.currentTimeMillis();
+			this.itemInfo = itemInfo;
+			this.workUnit = workUnit;
+		}
+	}
+
+	private static class StartedItem {
+		final String projectName;
+		final int buildNumber;
+
+		public StartedItem(final String projectName, final int buildNumber) {
+			this.projectName = projectName;
+			this.buildNumber = buildNumber;
+		}
+
+		@Override
+		public int hashCode() {
+			return Objects.hashCode(projectName, buildNumber);
+		}
+
+		@Override
+		public boolean equals(Object obj) {
+			if (obj == null)
+				return false;
+			if (getClass() != obj.getClass())
+				return false;
+			final StartedItem other = (StartedItem) obj;
+			return Objects.equal(this.projectName, other.projectName) && this.buildNumber == other.buildNumber;
```

Comment:
StartedItem gets instantiated for searching the right upstream item from UpstreamCauseStrategy#getPriority(). The upstream project name is taken from UpstreamCause there, which has explicit checks for null project names and IIRC I saw some null values there after a Jenkins restart, where downstream jobs were still in the queue. Sure, the check for null project names could be handled differently and more explicit, but it even works that way.